# Neighborhoods in Toronto
## Data Science Capstone Week 3

## Part 1 - Create the dataframe of Toronto Neighborhoods using *pandas*

I will begin by importing the *pandas* library, and then installing lxml to enable the read_html

In [2]:
import pandas as pd
pd.__version__

'1.0.5'

In [3]:
!pip install lxml

     |████████████████████████████████| 5.5MB 8.9MB/s eta 0:00:01


Set the Wikipedia page to object "url" and read it into a dataframe  
Check to see how many tables are created

In [4]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
df = pd.read_html(url)
print("df created: ", len(df) , "tables as type: ", type(df))

df created:  3 tables as type:  <class 'list'>


Seeing that the dataframe contains 3 tables, I only want to capture the first table (which contains the data necessary for this assignment)  
Reset "df" as the first index value

In [5]:
df = df[0]
df

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                          Neighborhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

Remove all rows where Borough is "Not assigned"...accomplished below by keeping only rows where Borough is not equal (!=) to "Not assigned"  
Reset the index

In [6]:
df = df[df.Borough != "Not assigned"]
df.reset_index(inplace = True)
df = df[["Postal Code", "Borough", "Neighborhood"]]
df.head(5)

Postal Code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

Show shape of the dataframe

In [7]:
df.shape

(103, 3)

## Part 2 - Get **latitude and longitude**

I will use the csv file provided in the instructions to create dataframe "df_ll"

In [8]:
latlong = "https://cocl.us/Geospatial_data"
df_ll = pd.read_csv(latlong)
df_ll.head(5)

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

Merge "df" with "df_ll", joining on the Postal Code column via a "left" join 

In [9]:
df_new = df.merge(df_ll, on = "Postal Code", how = "left")
df_new.head()

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

Check the shape of the merged dataframe "df_new"

In [10]:
df_new.shape

(103, 5)

## Part 3 - Explore the neighborhoods using *Foursquare* and cluster

Begin by bringing in libraries as seen in the previous assignment

In [18]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

# All requested packages already installed.

Solving environment: done

# All requested packages already installed.

Libraries imported.


For simplicity and per the assignment's suggestion, I will limit the dataframe to only neighborhoods which are in Boroughs containing the word "Toronto" 

In [23]:
toronto = df_new[df_new['Borough'].str.contains('Toronto')]
toronto.reset_index(inplace = True)
toronto = toronto[["Postal Code", "Borough", "Neighborhood", "Latitude", "Longitude"]]
print(toronto.shape)
toronto.head(5)

(39, 5)


Postal Code           Borough                                 Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.676357 -79.293031

Now the dataframe "toronto" has been limited to 39 postal-code neighborhoods

In [24]:
CLIENT_ID = '0CUBIUQXFQWAGKXPEADPCBLT2ZRQATMKB51WZTX1GUQDNKAE'
CLIENT_SECRET = 'GVZYF4JRMPU13MRGWPDKU4YIQQZZS4JQSJYYIGZR52VNY5RB'
VERSION = '20180624' #YYYMMDD today
LIMIT = 100 # limit of number of venues returned by Foursquare API


Create a get function to access *Foursquare* venue information for the different areas of Toronto

## **NOTE** that I am including the "Postal Code" in the get function...**I will be using "Postal Code" to group in many areas below, as it is easier to display than the list of Neighborhoods**

In [88]:
def getNearbyVenues(names, pcode, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, pc, lat, lng in zip(names, pcode, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            pc,
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood',
                  'Postal Code',
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)


toronto_venues = getNearbyVenues(names=toronto['Neighborhood'], pcode=toronto['Postal Code'], latitudes=toronto['Latitude'], longitudes=toronto['Longitude'])

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West, Forest Hill Road Park
High Park, The Junction South
North Toronto West, Lawrence Park
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
R

In [33]:
print(toronto_venues.shape)
print('Found {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))
toronto_venues.groupby("Postal Code").count()

(1620, 8)
Found 237 uniques categories.


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
Postal Code                                                                
M4E                     5                      5                       5   
M4K                    43                     43                      43   
M4L                    24                     24                      24   
M4M                    40                     40                      40   
M4N                     4                      4                       4   
M4P                     9                      9                       9   
M4R                    20                     20                      20   
M4S                    33                     33                      33   
M4T                     3                      3                       3   
M4V                    16                     16                      16   
M4W                     4                      4                       4   
M4X                    46                     46                      46   
M4Y                    75                     75                      75   
M5A                    44                     44                      44   
M5B                   100                    100                     100   
M5C                    78                     78                      78   
M5E                    56                     56                      56   
M5G                    64                     64                      64   
M5H                    92                     92                      92   
M5J                   100                    100                     100   
M5K                   100                    100                     100   
M5L                   100                    100                     100   
M5N                     1                      1                       1   
M5P                     6                      6                       6   
M5R                    21                     21                      21   
M5S                    34                     34                      34   
M5T                    62                     62                      62   
M5V                    16                     16                      16   
M5W                    96                     96                      96   
M5X                   100                    100                     100   
M6G                    17                     17                      17   
M6H                    16                     16                      16   
M6J                    43                     43                      43   
M6K                    24                     24                      24   
M6P                    25                     25                      25   
M6R                    15                     15                      15   
M6S                    38                     38                      38   
M7A                    34                     34                      34   
M7Y                    16                     16                      16   

             Venue  Venue Latitude  Venue Longitude  Venue Category  
Postal Code                                                          
M4E              5               5                5               5  
M4K             43              43               43              43  
M4L             24              24               24              24  
M4M             40              40               40              40  
M4N              4               4                4               4  
M4P              9               9                9               9  
M4R             20              20               20              20  
M4S             33              33               33              33  
M4T              3               3                3               3  
M4V             16              16               16              16  
M4W              4               4                4     

The function found 1620 venues across the Postal Codes, classified as 237 unique categories  

Next convert the categories to binary values using "get_dummies" function, and save as a new dataframe "toronto_onehot"  
Include Neighborhood and Postal Code in "toronto_onehot".  Since the value "Neighborhood" exists as a onehot column, define the new column as "Neighborhood_"

Then group the onehot data by Postal Code into a new dataframe "toronto_grouped"

In [90]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood_'] = toronto_venues['Neighborhood']
toronto_onehot['Postal Code'] = toronto_venues['Postal Code']


# move neighborhood and postal code columns to the beginning
fixed_columns = [toronto_onehot.columns[-1]] + [toronto_onehot.columns[-2]] + list(toronto_onehot.columns[:-2])
#fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

print(toronto_onehot.shape)
toronto_grouped = toronto_onehot.groupby('Postal Code').mean().reset_index()
print(toronto_grouped.shape)
toronto_grouped.head()

(1620, 239)
(39, 238)


Postal Code  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0         M4E                0.0      0.0                 0.0           0.0   
1         M4K                0.0      0.0                 0.0           0.0   
2         M4L                0.0      0.0                 0.0           0.0   
3         M4M                0.0      0.0                 0.0           0.0   
4         M4N                0.0      0.0                 0.0           0.0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0             0.0              0.0               0.0             0.000000   
1             0.0              0.0               0.0             0.023256   
2             0.0              0.0               0.0             0.000000   
3             0.0              0.0               0.0             0.050000   
4             0.0              0.0               0.0             0.000000   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0           0.0       0.0          0.0         0.0                  0.0   
1           0.0       0.0          0.0         0.0                  0.0   
2           0.0       0.0          0.0         0.0                  0.0   
3           0.0       0.0          0.0         0.0                  0.0   
4           0.0       0.0          0.0         0.0                  0.0   

   Asian Restaurant  Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  \
0               0.2                 0.0            0.0        0.0         0.0   
1               0.0                 0.0            0.0        0.0         0.0   
2               0.0                 0.0            0.0        0.0         0.0   
3               0.0                 0.0            0.0        0.0         0.0   
4               0.0                 0.0            0.0        0.0         0.0   

   Bagel Shop    Bakery   Bank    Bar  Baseball Stadium  Basketball Stadium  \
0         0.0  0.000000  0.000  0.000               0.0                 0.0   
1         0.0  0.023256  0.000  0.000               0.0                 0.0   
2         0.0  0.000000  0.000  0.000               0.0                 0.0   
3         0.0  0.050000  0.025  0.025               0.0                 0.0   
4         0.0  0.000000  0.000  0.000               0.0                 0.0   

   Beach  Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  \
0    0.0              0.0       0.0         0.0                 0.0   
1    0.0              0.0       0.0         0.0                 0.0   
2    0.0              0.0       0.0         0.0                 0.0   
3    0.0              0.0       0.0         0.0                 0.0   
4    0.0              0.0       0.0         0.0                 0.0   

   Bike Rental / Bike Share  Bistro  Board Shop  Boat or Ferry  Bookstore  \
0                       0.0     0.0    0.000000            0.0   0.000000   
1                       0.0     0.0    0.000000            0.0   0.046512   
2                       0.0     0.0    0.041667            0.0   0.000000   
3                       0.0     0.0    0.000000            0.0   0.025000   
4                       0.0     0.0    0.000000            0.0   0.000000   

   Boutique  Brazilian Restaurant  Breakfast Spot   Brewery  Bubble Tea Shop  \
0       0.0                   0.0             0.0  0.000000         0.000000   
1       0.0                   0.0             0.0  0.023256         0.023256   
2       0.0                   0.0             0.0  0.041667         0.000000   
3       0.0                   0.0             0.0  0.050000         0.000000   
4       0.0                   0.0             0.0  0.000000         0.000000   

   Building  Burger Joint  Burrito Place  Bus Line  Butcher  Cafeteria  \
0       0.0           0.0       0.000000      0.00      0.0        0.0   
1       0.0           0.0       0.000000      0.00      0.0        0.0   
2       0.0           0.0       0.041667      0.00      0.

Create *pandas* dataframe with the top 10 venue categories per Postal Code Neighborhood

In [53]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

num_top_venues = 10
indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Postal Code']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
postal_code_venues_sorted = pd.DataFrame(columns=columns)
postal_code_venues_sorted['Postal Code'] = toronto_grouped['Postal Code']

for ind in np.arange(toronto_grouped.shape[0]):
    postal_code_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

postal_code_venues_sorted.head()

Postal Code 1st Most Common Venue 2nd Most Common Venue  \
0         M4E     Health Food Store          Neighborhood   
1         M4K      Greek Restaurant           Coffee Shop   
2         M4L                  Park  Fast Food Restaurant   
3         M4M                  Café           Coffee Shop   
4         M4N                  Park              Bus Line   

  3rd Most Common Venue 4th Most Common Venue   5th Most Common Venue  \
0      Asian Restaurant                   Pub                   Trail   
1    Italian Restaurant        Ice Cream Shop  Furniture / Home Store   
2           Pizza Place            Board Shop       Fish & Chips Shop   
3             Gastropub                Bakery                 Brewery   
4           Swim School                Lawyer    Colombian Restaurant   

  6th Most Common Venue        7th Most Common Venue 8th Most Common Venue  \
0      Department Store                 Dessert Shop    Dim Sum Restaurant   
1            Restaurant                    Bookstore       Bubble Tea Shop   
2               Brewery                 Liquor Store            Restaurant   
3   American Restaurant                 Neighborhood        Sandwich Place   
4           College Gym  Eastern European Restaurant   Dumpling Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0                 Diner            Yoga Studio  
1     Indian Restaurant                    Pub  
2         Burrito Place     Italian Restaurant  
3           Cheese Shop              Pet Store  
4            Donut Shop       Doner Restaurant

Begin clustering with k-means method, setting k = 5

In [77]:
# set number of clusters
k = 5

toronto_grouped_clustering = toronto_grouped.drop('Postal Code', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 1, 1, 1, 4, 1, 1, 1, 3, 1], dtype=int32)

In [80]:
# add clustering labels
postal_code_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)
#use below (instead of above line) if updating k & "Cluster Labels" has already been inserted
#postal_code_venues_sorted["Cluster Labels"] = kmeans.labels_
toronto_merged = toronto

# merge "toronto_grouped" with "toronto" to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.merge(postal_code_venues_sorted.set_index('Postal Code'), on = "Postal Code", how = "left")
toronto_merged.head()

Postal Code           Borough                                 Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636               1           Coffee Shop   
1  43.662301 -79.389494               1           Coffee Shop   
2  43.657162 -79.378937               1        Clothing Store   
3  43.651494 -79.375418               1                  Café   
4  43.676357 -79.293031               1     Health Food Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Park                   Pub                Bakery   
1      Sushi Restaurant                 Diner           Yoga Studio   
2           Coffee Shop                  Café        Cosmetics Shop   
3           Coffee Shop             Gastropub            Restaurant   
4          Neighborhood      Asian Restaurant                   Pub   

       5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                       Café               Theater            Restaurant   
1             Discount Store                   Bar              Beer Bar   
2  Middle Eastern Restaurant   Japanese Restaurant       Bubble Tea Shop   
3        American Restaurant          Cocktail Bar        Lingerie Store   
4                      Trail      Department Store          Dessert Shop   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0        Breakfast Spot           Yoga Studio             Shoe Store  
1         Smoothie Shop      Sculpture Garden         Sandwich Place  
2    Italian Restaurant               Theater       Ramen Restaurant  
3   Moroccan Restaurant              Creperie                  Hotel  
4    Dim Sum Restaurant                 Diner            Yoga Studio

Map the clusters using a Folium map  


In [83]:
latitude = 43.6532
longitude = -79.3832
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
#colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
#rainbow = [colors.rgb2hex(i) for i in colors_array]
rainbow = ["purple", "blue", "red", "green", "yellow"]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Postal Code'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters